diff --git a/CMakeLists.txt b/CMakeLists.txt
index 256536f9..ff08e086 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -68,6 +68,11 @@ set(CMAKE_POSITION_INDEPENDENT_CODE YES)
 
 # Fetchable dependcies
 include(FetchContent)
+if (_SwiftSyntax_SourceDIR)
+    FetchContent_Declare(SwiftSyntax
+        SOURCE_DIR ${_SwiftSyntax_SourceDIR})
+    message(STATUS "_SwiftSyntax_SourceDIR: ${_SwiftSyntax_SourceDIR}")
+endif()
 if (_SwiftFoundationICU_SourceDIR)
     FetchContent_Declare(SwiftFoundationICU
         SOURCE_DIR ${_SwiftFoundationICU_SourceDIR})
