From 48a2e57c77351aeaa02e544772970d2d3f8515e1 Mon Sep 17 00:00:00 2001
From: Andrew Druk <adruk@readdle.com>
Date: Mon, 23 Dec 2024 02:35:54 +0200
Subject: [PATCH] Dev: fix linux/stat header in ForSwiftFoundationOnly.h

---
 Sources/CoreFoundation/include/ForSwiftFoundationOnly.h | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/Sources/CoreFoundation/include/ForSwiftFoundationOnly.h b/Sources/CoreFoundation/include/ForSwiftFoundationOnly.h
index 88fb3d36..53a5d560 100644
--- a/Sources/CoreFoundation/include/ForSwiftFoundationOnly.h
+++ b/Sources/CoreFoundation/include/ForSwiftFoundationOnly.h
@@ -69,6 +69,13 @@
 #include <sys/stat.h>
 #include <sys/syscall.h>
 #include <termios.h>
+#include <linux/fcntl.h>
+#ifdef __swift__
+// The linux/stat header is private in the Android modulemap.
+#pragma clang module import posix_filesystem.linux_stat
+#else
+#include <linux/stat.h>
+#endif
 #elif TARGET_OS_WASI
 #include <fcntl.h>
 #include <sys/stat.h>
-- 
2.46.0

