From 49dac2443c37c2bff59c9294c096b2948cace43b Mon Sep 17 00:00:00 2001
From: Anton Pogonets <apogonets@readdle.com>
Date: Fri, 9 Apr 2021 13:38:14 +0300
Subject: [PATCH] Dont use posix_spawnattr on Android. It available only from
 API>=28

---
 Sources/Foundation/Process.swift | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/Sources/Foundation/Process.swift b/Sources/Foundation/Process.swift
index 7d6a1a39..cfe4cd19 100644
--- a/Sources/Foundation/Process.swift
+++ b/Sources/Foundation/Process.swift
@@ -950,6 +950,8 @@ open class Process: NSObject, @unchecked Sendable {
             ])
         }
 #endif
+
+#if !os(Android)
         try _throwIfPosixError(posix_spawnattr_init(&spawnAttrs))
         try _throwIfPosixError(posix_spawnattr_setflags(&spawnAttrs, .init(POSIX_SPAWN_SETPGROUP)))
 #if canImport(Darwin)
@@ -964,6 +966,7 @@ open class Process: NSObject, @unchecked Sendable {
             }
             try _throwIfPosixError(_CFPosixSpawnFileActionsAddClose(fileActions, fd))
         }
+#endif
 #endif
 
         // Unsafe fallback for systems missing posix_spawn_file_actions_addchdir[_np]
@@ -995,8 +998,9 @@ open class Process: NSObject, @unchecked Sendable {
                 throw _NSErrorWithErrno(errno, reading: true, path: launchPath)
             }
         })
+#if !os(Android)
         posix_spawnattr_destroy(&spawnAttrs)
-
+#endif
         // Close the write end of the input and output pipes.
         if let pipe = standardInput as? Pipe {
             pipe.fileHandleForReading.closeFile()
-- 
2.46.0

