From 5634e39e22b102d27a70b98b2579d599090f4d8c Mon Sep 17 00:00:00 2001
From: Andrew Druk <adruk@readdle.com>
Date: Mon, 14 Apr 2025 17:17:58 +0300
Subject: [PATCH] Add SwiftSyntax source dir variable

---
 CMakeLists.txt | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index c0d46e43..3c611c97 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -83,6 +83,11 @@ set(CMAKE_POSITION_INDEPENDENT_CODE YES)
 
 # Fetchable dependencies
 include(FetchContent)
+if (_SwiftSyntax_SourceDIR)
+    FetchContent_Declare(SwiftSyntax
+        SOURCE_DIR ${_SwiftSyntax_SourceDIR})
+    message(STATUS "_SwiftSyntax_SourceDIR: ${_SwiftSyntax_SourceDIR}")
+endif()
 if (_SwiftFoundationICU_SourceDIR)
     FetchContent_Declare(SwiftFoundationICU
         SOURCE_DIR ${_SwiftFoundationICU_SourceDIR})
-- 
2.46.0

