diff --git a/Sources/SwiftLibraryPluginProvider/LibraryPluginProvider.swift b/Sources/SwiftLibraryPluginProvider/LibraryPluginProvider.swift
index eb06cca2..284d117d 100644
--- a/Sources/SwiftLibraryPluginProvider/LibraryPluginProvider.swift
+++ b/Sources/SwiftLibraryPluginProvider/LibraryPluginProvider.swift
@@ -20,6 +20,8 @@ private import _SwiftLibraryPluginProviderCShims
 private import Darwin
 #elseif canImport(Glibc)
 private import Glibc
+#elseif canImport(Android)
+private import Android
 #elseif canImport(Musl)
 private import Musl
 #endif
@@ -31,6 +33,8 @@ import SwiftSyntaxMacros
 @_implementationOnly import Darwin
 #elseif canImport(Glibc)
 @_implementationOnly import Glibc
+#elseif canImport(Android)
+@_implementationOnly import Android
 #elseif canImport(Musl)
 @_implementationOnly import Musl
 #endif
@@ -137,7 +141,12 @@ private func _loadLibrary(_ path: String) throws -> UnsafeMutableRawPointer {
 #else
 private func _loadLibrary(_ path: String) throws -> UnsafeMutableRawPointer {
   guard let dlHandle = dlopen(path, RTLD_LAZY | RTLD_LOCAL) else {
-    throw LibraryPluginError(message: "loader error: \(String(cString: dlerror()))")
+    #if canImport(Android)
+      let err = String(cString: dlerror()!)
+    #else
+      let err = String(cString: dlerror())
+    #endif
+    throw LibraryPluginError(message: "loader error: \(err)")
   }
   return dlHandle
 }
