From 83574307e1f3aa5f67da93f4d93dc58f3bffe0fc Mon Sep 17 00:00:00 2001
From: Andrew Druk <adruk@readdle.com>
Date: Sat, 12 Apr 2025 15:19:51 +0300
Subject: [PATCH] Add presets for android x86 x86_64

---
 utils/build-presets.ini | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/utils/build-presets.ini b/utils/build-presets.ini
index 14b9a7006d2..23abcd95d91 100644
--- a/utils/build-presets.ini
+++ b/utils/build-presets.ini
@@ -1026,6 +1026,20 @@ mixin-preset=buildbot_linux_crosscompile_android,tools=RA,stdlib=RD,build
 
 android-arch=aarch64
 
+[preset: buildbot_linux_crosscompile_android,tools=RA,stdlib=RD,build,x86]
+mixin-preset=buildbot_linux_crosscompile_android,tools=RA,stdlib=RD,build
+
+dash-dash
+
+android-arch=i686
+
+[preset: buildbot_linux_crosscompile_android,tools=RA,stdlib=RD,build,x86_64]
+mixin-preset=buildbot_linux_crosscompile_android,tools=RA,stdlib=RD,build
+
+dash-dash
+
+android-arch=x86_64
+
 # Ubuntu 18.04 preset for backwards compat and future customizations.
 [preset: buildbot_linux_1804]
 mixin-preset=buildbot_linux
-- 
2.46.0

