diff --git a/utils/build-presets.ini b/utils/build-presets.ini
index 66665164c9b..7c57361b97b 100644
--- a/utils/build-presets.ini
+++ b/utils/build-presets.ini
@@ -862,8 +862,6 @@ build-swift-stdlib-unittest-extra
 # Executes the lit tests for the installable package that is created
 # Assumes the swift-integration-tests repo is checked out
 
-test-installable-package
-
 # Build the benchmarks against the toolchain.
 toolchain-benchmarks
 
@@ -970,12 +968,6 @@ mixin-preset=
 build-subdir=buildbot_linux
 
 release
-test
-validation-test
-long-test
-stress-test
-test-optimized
-lit-args=-v --time-tests
 
 build-ninja
 
@@ -992,8 +984,6 @@ install-destdir=%(install_destdir)s
 # Path to the .tar.gz package we would create.
 installable-package=%(installable_package)s
 
-host-test
-
 install-prefix=/usr
 install-llvm
 install-static-linux-config
