diff --git a/lib/IRGen/IRGenModule.cpp b/lib/IRGen/IRGenModule.cpp
index f647c38a2e3..eda1923fd99 100644
--- a/lib/IRGen/IRGenModule.cpp
+++ b/lib/IRGen/IRGenModule.cpp
@@ -1740,7 +1740,7 @@ bool IRGenModule::useDllStorage() { return ::useDllStorage(Triple); }
 bool IRGenModule::shouldPrespecializeGenericMetadata() {
   auto canPrespecializeTarget =
       (Triple.isOSDarwin() ||
-       (Triple.isOSLinux() && !(Triple.isARM() && Triple.isArch32Bit())));
+       (Triple.isOSLinux() && !(Triple.isArch32Bit())));
   if (canPrespecializeTarget && isStandardLibrary()) {
     return IRGen.Opts.PrespecializeGenericMetadata;
   }
