From c49c13cd47983030cf6b880ba18317203c629c5e Mon Sep 17 00:00:00 2001
From: Anton Pogonets <apogonets@readdle.com>
Date: Wed, 21 Nov 2018 19:31:39 +0200
Subject: [PATCH] Add LIBDISPATCH_DEFAULT_THREAD_POOL_SIZE env variable

---
 src/internal.h |  8 +++++++-
 src/queue.c    | 11 ++++++++++-
 2 files changed, 17 insertions(+), 2 deletions(-)

diff --git a/src/internal.h b/src/internal.h
index 3cc16fc..da0e36a 100644
--- a/src/internal.h
+++ b/src/internal.h
@@ -1112,6 +1112,12 @@ extern bool _dispatch_kevent_workqueue_enabled;
 #include "inline_internal.h"
 #include "firehose/firehose_internal.h"
 
-__END_DECLS
+#if !HAVE_PTHREAD_WORKQUEUE_KEVENT
+// copied from https://opensource.apple.com/source/libpthread/libpthread-301.50.1/kern/workqueue_internal.h.auto.html
+#define WORKQUEUE_MAXTHREADS 512
+#define WORKQUEUE_CONSTRAINED_MAXTHREADS (WORKQUEUE_MAXTHREADS >> 3)
+#define WORKQUEUE_CONSTRAINED_FACTOR 5
+#endif
 
+__END_DECLS
 #endif /* __DISPATCH_INTERNAL__ */
diff --git a/src/queue.c b/src/queue.c
index 90f3cfa..9537d6f 100644
--- a/src/queue.c
+++ b/src/queue.c
@@ -6168,10 +6168,19 @@ _dispatch_root_queue_init_pthread_pool(dispatch_queue_global_t dq,
 		int pool_size, dispatch_priority_t pri)
 {
 	dispatch_pthread_root_queue_context_t pqc = dq->do_ctxt;
+	int32_t default_pool_size = 0;
+	char* default_pool_size_env = getenv("LIBDISPATCH_DEFAULT_THREAD_POOL_SIZE");
+	if (default_pool_size_env) {
+		default_pool_size = (int32_t) atoi(default_pool_size_env);
+	}
+	if (!default_pool_size) {
+		default_pool_size = (int32_t) MAX(dispatch_hw_config(active_cpus) * 5, WORKQUEUE_CONSTRAINED_MAXTHREADS);
+	}
 	int thread_pool_size = DISPATCH_WORKQ_MAX_PTHREAD_COUNT;
 	if (!(pri & DISPATCH_PRIORITY_FLAG_OVERCOMMIT)) {
-		thread_pool_size = (int32_t)dispatch_hw_config(active_cpus);
+		thread_pool_size = default_pool_size;
 	}
+
 	if (pool_size && pool_size < thread_pool_size) thread_pool_size = pool_size;
 	dq->dgq_thread_pool_size = thread_pool_size;
 	qos_class_t cls = _dispatch_qos_to_qos_class(_dispatch_priority_qos(pri) ?:
-- 
2.29.2

